text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,308 @@
+
+```
+S = EachOf(
+     AnyOf(
+       ConsistentParentBased(
+         RuleBased(ROOT, {
+             (http.target == /healthcheck) => ConsistentAlwaysOff,
+             (http.target == /checkout) => ConsistentAlwaysOn,
+             true => ConsistentFixedThreshold(0.25)
+         }),
+       RuleBased(CLIENT, {
+         (http.url == /foo) => ConsistentAlwaysOn
+       }
+     ),
+     ConsistentRateLimiting(1000 * 60)
+   )
+```
+
+### Limitations of composite samplers in Approach One
+
+Not all samplers can participate as components of composite samplers without undesired or unexpected effects. Some samplers require that they _see_ each `Span` being created, even if the span is going to be dropped. Some samplers update the trace state or maintain internal state, and for their correct behavior it it is assumed that their sampling decisions will be honored by the tracer at the face value in all cases. A good example for this are rate limiting samplers which have to keep track of the rate of created spans and/or the rate of positive sampling decisions.
+
+A special attention is required for CP samplers. The sampling probability they record in trace-state is later used to calculate [_adjusted count_](https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling/#adjusted-count), which, in turn, is used to calculate span-based metrics. While the composite samplers presented here are compatible with CP samplers, generally, mixing CP samplers with other types of samplers may lead to undefined or sometimes incorrect adjusted counts.
+
+## Approach Two
+
+A principle of operation for Approach Two is that `ShouldSample` is invoked only once, on the root of the tree formed by composite samplers. All the logic provided by the composition of samplers is handled by calculating the threshold values, delegating the calculation downstream as necessary.
+
+### Consistent Probability Sampler API
+
+To make this approach possible, all Consistent Probability Samplers need to implement the following API, in addition to the standard Sampler API. This extension will be used by all composite samplers in Approach Two, as listed in the next sections.
+
+#### GetThreshold
+
+This is a routine/function/method for all Consistent Probability Samplers. Its purpose is to query the sampler to provide the rejection threshold value they would use had they been asked to make a sampling decision for a given span, however, without constructing the actual sampling Decision.
+
+#### Required Arguments for GetThreshold:
+
+The arguments are the same as for [`ShouldSample`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shouldsample).
+
+- `Context` with parent Span.
+- `TraceId` of the Span to be created.
```

Comment:
I think `TraceId` should not be an argument as the threshold must not depend on the randomness. By providing the trace ID, there is some risk that the lower 56-bits will be (mis-)used for the threshold computation.
```suggestion
```

Reply:
Great suggestion, thanks!